docs: add frontend UI and usability guide (#999)
ElektraInitiative/PermaplanT: doc/guidelines/frontend-ui-usability.md

# Frontend UI & Usability

## Terminology

### Usability

> Extent to which a system, product or service can be used by specified users to achieve specified goals with effectiveness, efficiency and satisfaction in a specified context of use.

### User Interface (UI)

> Set of all the components of an interactive system that provide information and controls for the user to accomplish specific tasks with the interactive system.

### User Experience (UX)

> Combination of user's perceptions and responses that result from the use and/or anticipated use of a system, product or service.

## Guidelines

This guide should help improve the UI (User Interface) and usability of the PermaplanT web application going forward.  
The following guidelines are:

- on the one hand a mix of general UI and usability best practices and
- on the other hand design suggestions tailored to PermaplanT.

### Text

- **Alignment**

>   - Left align when text spans over more than three lines because centered text becomes hard to follow when having to jump to the next line multiple times.

Reviewer: Formulate everything as rule if possible. So is here the rule that we should be below three lines if possible? Which alignment to choose if we are above 3 lines?

  - Avoid mixing two different alignments in a text section, e.g. a centered headline should be followed by centered text.

- **Spacing / margin**  

>   Multiplier strategy to calculate the margins - if in doubt, use a multiplier of 2, e.g.:

Reviewer: Can we give concrete guidelines which px to use?

```suggestion
  Multiplier strategy to calculate the margins - if in doubt, use a multiplier of 2. Generally we use:
```

  - Text sections:  
    headline-1 <- 16px -> text <- 32px -> headline-2
  - Navigation bar:  
    logo <- 48px -> navigation-link-1 <- 24px -> navigation-link-2 <- 24px -> navigation-link-3

- **Letter spacing**  

>   Often times, like in PermaplanT's current case, the default value (in Tailwind: `tracking-normal`) is good enough.

Reviewer: Is the list below exceptions?

```suggestion
  Often times, like in PermaplanT's current case, the default value (in Tailwind: `tracking-normal`) is good enough.
  Exceptions:
```

  - Headings: negative values can be used for the bigger headings h1-h3, i.e. `tracking-tighter` or `tracking-tight`
  - Running text: usually can be left as is but, if needed, can be set to `tracking-wide` or `tracking-wider`

### Colors

PermaplanT uses following **three color palettes**:

> - **Neutral** ~60% (PermaplanT: gray): **dominant** color of the design, e.g. used for background, text and labels

Reviewer: What means the 60%? How much it is used?

- If **textual links**, e.g. the links in PermaplanT's navigation bar, need to be surrounded by space, always set their `margin` instead of `padding`.
- Make **call-to-action buttons** look the same everywhere by using our `SimpleButton` component.
  Use Tailwind's classes to define the button's `margin` and `width`, if necessary due to layout or viewport.
- **Toast errors** are always shown on the top right.
  To achieve this, we are simply using the default setting of our used library (_react-toastify_).
- Use the same **application-specific terms** everywhere when communicating to the user, e.g. always write _PermaplanT_.
- **Forms** and their fields should look and behave the same everywhere according to our guidelines in [Forms](#forms).
- In every **form**, **validation** errors are detected and shown by using the native input validation of HTML5.

### Icons

- In their **passive**, i.e. currently not active/enabled, state they must appear in **neutral** colors/colorless.
- When **active**, i.e. currently activated/enabled, they must get a small **highlighting** through a visually stronger and more colorful appearance by using the design's primary colors.

> - When **disabled**, i.e. currently not clickable, they must be **greyed-out**, either via decreasing its opacity or choosing light, faint colors.  

Reviewer: How much decreasing? Why having two options?

  The mouse cursor must be styled with the `not-allowed` css property.
- A **tooltip** on hovering must be displayed for every icon in PermaplanT's toolbox.

>   The tooltip contains the icon's **label** (as concise as possible) and, if existing, the assigned **shortcut**.

Reviewer: Preferably we should order the guidelines by priority. E.g. isn't this more important then the colors for passive/active/disabled?

Reviewer 2: I agree. Maybe also with a table of contents, so someone can quickly find what they are searching.

### Highlighting

Highlighting techniques:

- Font weight:

  - **Bold** text adds tolerable noise to the design and enables distinct highlighting in both short and long texts.
    In general, it is the preferred highlighting technique to use.
  - _Italic_ text adds minimal noise to the design, but lacks in recognizability.
  - <u>Underlining</u> adds most noise and compromises a text's legibility.
    It should not be used.

- Colors: primary and secondary colors of the design's color palette

- Images: users generally remember images better than words

> - Avoid different fonts for highlighting purposes.

Reviewer: Are different fonts allowed for any purpose?

# Frontend UI & Usability

> ## Terminology

Reviewer 2: Maybe the terminology is better suited inside our glossary.

### User Experience (UX)

> Combination of user's perceptions and responses that result from the use and/or anticipated use of a system, product or service.

## Guidelines

This guide should help improve the UI (User Interface) and usability of the PermaplanT web application going forward.  
The following guidelines are:

- on the one hand a mix of general UI and usability best practices and
- on the other hand design suggestions tailored to PermaplanT.

### Text

- **Size**

  - `<h1>` Headings : `text-2xl`
  - `<h2>` Subheadings: `text-xl`
  - `<h3>` Subheadings (currently not used/needed): `text-lg`
  - Navigation items: `text-lg`
  - Running text: `text-sm`
  - Form fields and labels: `text-sm`

- **Alignment**

  - Always left align everything.
  - Exceptions:
    - Modals: center headings
    - Guided tour: center highlighted actions

- **Spacing / margin**

  - `<h1>` Headings: `mb-8`
  - `<h2>` Subheadings: `mb-2`
  - Cards (like on our Map overview): `mb-4`
  - Form fields (like on the Seeds page): `mb-8`
  - Labels of form fields: `mb-2`

>   - If in doubt, use the multiplier strategy to calculate margins between sections, e.g. we are using a multiplier of 4 in our navigation bar:  

Reviewer 3: It is unclear to me what the multiplier strategy is.

    logo <- 48px -> navigation-link-1 <- 24px -> navigation-link-2 <- 24px -> navigation-link-3

- **Letter spacing**

  - Always go with the default value, i.e. `tracking-normal`.

>   - In case letter spacing changes are considered in the future:

Reviewer 3: I don't think we need considerations for the future in this document.
It it is more effort to read through the guideline and it doesn't apply anyways.

Reviewer 3: I agree with the statement though :D

Reviewer: We could move future guidelines to a final section?

  - Always go with the default value, i.e. `tracking-normal`.
  - In case letter spacing changes are considered in the future:
    - Headings: use negative values for the bigger headings h1-h3, i.e. `tracking-tighter` or `tracking-tight`
    - Running text: `tracking-wide` or `tracking-wider`

> ### Colors

Reviewer 2: Which color should be used for cancel buttons e.g (cancel tour, cancel map/seed creation)?

- **Neutral** (PermaplanT: gray): **dominant** color of the design, e.g. used for background, text and labels.  
  Optimal usage: ~60%
- **Primary** (PermaplanT: asparagus green): PermaplanT's **main brand** color, e.g. used for buttons, hovering, highlighting focused input fields.  

>   Optimal usage: ~30%

Reviewer 3: I don't agree with the percentages. I don't think they are helpful.

According to the material guideline the primary color should be used for the most prominent/important actions and therefore might be used less than the secondary color.

- **Secondary** (PermaplanT: sea blue): PermaplanT's **second brand** color, e.g. used for highlighting a few UI elements (e.g. in the layers section) and action-texts.  

>   Optimal usage: ~10%

Reviewer 3: Please add some examples on when to use the main brand vs the second brand color.

Check out Google's [Material Design Guidelines](https://m3.material.io/styles/color/the-color-system/key-colors-tones) for more information on color palettes and their usage.

> **Only those** three color palettes **should be used** throughout the whole application.

Reviewer 3: I disagree with this statement. It occasionally makes sense to use other colors which have some sort of association. E.g. usage of the color red to indicate errors or destructive actions.

Reviewer: Then we probably should define more colors, e.g. the colors used for the companions? But I agree with @tschawax  that we shouldn't leave it to the respective devs to bring arbitrary other colors into the app as wanted. If anywhere the color for e.g. the companions is needed again, it should be the same color as already used.

> - Tailwind's default color palette for **red** is e.g. used to render validation errors in forms, form field asterisks, buttons for destructive actions.

Reviewer 4: maybe we could mention the exact color that should be user for the different use cases. For example you mentioned below that the astrisks should be red-500, but i would excpect this here in the color section.

| main color | 500 | 300 |
| text on main | 50 | 700 |
| alternative color | 200 | 600 |
| text on alternative | 800 | 200 |

### Forms

#### Design

- Display a **left aligned heading** if the form is representing a whole page, e.g. PermaplanT's forms to create maps and seeds.
- Add a **capitalized label** above each field.
- **Align** labels to the **left** of their corresponding fields.
- Always use the **_for_** HTML attribute (in React: _htmlFor_) to bind the label to the field.
- Always use PermaplanT's **SearchInput** component to declare **search fields** instead of declaring them with type _search_.
- **Don't use** the **title** attribute for any form fields (see <https://inclusive-components.design/tooltips-toggletips/>)
- All form fields should have the **same styling**:
  - border color: `border-neutral-500 dark:border-neutral-400-dark`
  - border focus: `focus:border-primary-500 dark:focus:border-primary-300`
  - border radius: `rounded-lg`
  - padding: `p-2.5`
  - placeholder: `placeholder-neutral-300`
  - text: `text-sm`

#### Error Prevention

- Use the **corresponding _type_** of the input field, e.g. _date_ for date fields, _number_ for numerical-only inputs etc.

> - Mark required fields with an **asterisk**:

Reviewer 2: The asterisk color is already mentioned in colors.

  - color: `text-red-500`
  - all other styles are inherited from the parent label
- Use HTML-native **constraint** attributes for input fields where appropriate, e.g. _maxlength_, _required_, _step_, etc.
- Use short **placeholder** texts for input fields to show allowed values or further explain the intent of the input field.
- Users should **never** be **required** to manually enter a **metrical** unit or a **currency** symbol.
  Instead, make the unit part of the input field itself or leave it out if it's clear from the context.
- **Submission** to the backend should not happen until all fields are **verified** by the **frontend** logic.

> - In case of submissions **on-the-fly**, e.g. setting dates for plants, the data should always be submitted with **debouncing**.

Reviewer 2: Debouncing should be a general idea (e.g for search bars), not only for forms.

- **Double submission** should be **prevented** by disabling the submission button until the server has responded.

### Consistency

> Things with the **same meaning** should **look and behave** the **same** throughout the whole application.

Reviewer 2: Which "things"? Words and actions?

- **Headings** of all toolbars are always defined by `<h2>`
- If **textual links**, e.g. the links in PermaplanT's navigation bar, need to be surrounded by space, always set their `margin` instead of `padding`.
- Make **call-to-action buttons** look the same everywhere by using our `SimpleButton` component.
  Use Tailwind's classes to define the button's `margin` and `width`, if necessary due to layout or viewport.
- **Toast errors** are always shown on the top right.
  To achieve this, we are simply using the default setting of our used library (_react-toastify_).
- Use the same **application-specific terms** everywhere when communicating to the user, e.g. always write _PermaplanT_.
- **Forms** and their fields should look and behave the same everywhere according to our guidelines in [Forms](#forms).
- In every **form**, **validation** errors are detected and shown by using the native input validation of HTML5.

### Icons

- A **tooltip** on hovering must be displayed for every icon in PermaplanT's toolbox.
  The tooltip contains the icon's **label** (as concise as possible) and, if existing, the assigned **shortcut**.

> - In their **passive**, i.e. currently not active/enabled, state they must appear in **neutral** colors/colorless.

Reviewer 3: ```suggestion
- In their **passive**, i.e. currently not active/enabled state they must appear in **neutral** colors/colorless.
```

- When **active**, i.e. currently activated/enabled, they must get a small **highlighting** through a visually stronger and more colorful appearance by using the design's primary colors.
- When **disabled**, i.e. currently not clickable, they must be **greyed-out**, via decreasing their opacity.  
  The mouse cursor must be styled with the `not-allowed` css property.

### Highlighting

Highlighting techniques:

- Font weight:

  - **Bold** text adds tolerable noise to the design and enables distinct highlighting in both short and long texts.
    In general, it is the preferred highlighting technique to use.
  - _Italic_ text adds minimal noise to the design, but lacks in recognizability.
  - <u>Underlining</u> adds most noise and compromises a text's legibility.
    It should not be used.

- Colors: primary and secondary colors of the design's color palette

- Images: users generally remember images better than words

> - Never use a different font family.

Reviewer 3: Which font-family is / should be used?
Is there one defined? Do we want one or fallback to the system font of the user?
Do we want to have a different font-family for headlines?

> - Colors: primary and secondary colors of the design's color palette

Reviewer 2: I personally think blue is not the best color for text highlighting, because of the resemblance to hyperlinks.

### Wording

Messages shown to the user should strive to fulfill following criteria:

- **concise**: the more text the less likely it will be read by the user
- **clear**: straight to the message's essence
- **understandable** for technical laymen: no status codes or technical terms
- **no exclamation** marks: interpreted as commanding
- **no uppercase** words: comes across as shouting
- **detailed** information should be hidden behind a **read more** link or a **collapsed** section
- **Headlines**, **sub-headlines** and **labels** should be **capitalized**
- **Minimize hyphenation** of words

> - **American English** spelling

Reviewer 3: and german

Reviewer: ```suggestion
- **American English** and **German** (Neue Rechtschreibung) spelling
```
